lib/teleterm/gateway/local_proxy_middleware.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gateway
+
+import (
+	"context"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	alpn "github.com/gravitational/teleport/lib/srv/alpnproxy"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+type localProxyMiddleware struct {
+	onExpiredCert func(context.Context) error
+	log           *logrus.Entry
+	dbRoute       tlsca.RouteToDatabase
+}
+
+// OnNewConnection calls m.onExpiredCert if the cert used by the local proxy has expired.
+// This is a very basic reimplementation of client.DBCertChecker.OnNewConnection. DBCertChecker
+// supports per-session MFA while for now Connect needs to just check for expired certs.
+//
+// In the future, DBCertChecker is going to be extended so that it's used by both tsh and Connect
+// and this middleware will be removed.
+func (m *localProxyMiddleware) OnNewConnection(ctx context.Context, lp *alpn.LocalProxy, conn net.Conn) error {
+	m.log.Debug("Checking local proxy certs")
+
+	err := lp.CheckDBCerts(m.dbRoute)
```

Comment:
The CheckDBCerts seems to also returns diffrents errors apart from `x509.CertificateInvalidError` like: `certificate subject is for user` Should we call onExpireCert login in case of "certificate subject is for user" error ?

Reply:
> The CheckDBCerts seems to also returns diffrents errors apart from `x509.CertificateInvalidError` like: `certificate subject is for user` Should we call onExpireCert login in case of "certificate subject is for user" error ?

We shouldn't, I'll post a fix in a sec which inspects the error more closely.

Reply:
CheckDBCerts looks at db name/user in the database route in case user has previously logged in with `tsh db login` using a different db name/user than what they used for `tsh proxy db --tunnel --db-user=$dbuser --db-name=$dbname`, and the previous cert was loaded from disk

For teleterm flow, why should a database route mismatch not trigger cert reissue?

Reply:
Good question, I went back and forth on this when writing this reply.

In general, Connect doesn't let you execute the equivalent of `tsh db login` and `tsh proxy db --tunnel` separately. Both always happen within a single UI interaction of creating the gateway so you cannot provide different args to them. During gateway creation, we obtain a lock on `daemon.Service.mu`. This way we guarantee that no two gateways are being created at the same time… but only within Connect. The cert for the gateway is fetched right before we create it, but unlike `tsh proxy db` we save and load it from disk, even during reissue managed by the middleware.

And I guess here's where problems might happen. Technically we cannot guarantee that between being saved and loaded the cert is not modified by, let's say, `tsh db login`. Connect also has no real need to save the cert to disk, this is just the tech debt of not supporting per-session MFA: Connect calls `TeleportClient.ReissueUserCerts` (which doesn't return the cert) rather than `TeleportClient.IssueUserCertsWithMFA`.

This quarter we also plan to make Connect use `~/.tsh` rather than its own `TELEPORT_HOME` so the risk of some weird mismatched will be increased.

From what I see, the db cert on disk is scoped per cluster user & db server name, so I suppose we could run into something like this:

```mermaid
sequenceDiagram
    participant connect as Connect
    participant keys as key agent
    participant tsh as outside tsh
    connect ->> keys: Save db cert for db user foo
    tsh ->> keys: Save db cert for db user bar
    connect ->> keys: Get cert for db
    keys ->> connect: Return cert for db user bar
    connect ->> connect: Start local proxy
```

Which means:

* The mismatch between the cert subject and the local proxy params won't happen in Connect unless an incorrect cert was loaded into the gateway.
* For now, Connect cannot avoid loading certs from the disk.
* …so Connect should `CheckDBCerts` before instantiating `alpn.LocalProxy` and before doing `alpn.LocalProxy.SetCerts` when reissuing certs and return an error if there's a mismatch.
* By always setting the cert only after we check it, we ensure that in Connect, `OnNewConnection` cannot encounter a cert subject mismatch.
* …so Connect should treat that as unrecoverable error and not execute `OnExpiredCert`.

@GavinFrazar So I think the way to proceed would be to keep current code, that is Connect's middleware will not call `OnExpiredCert` if the check failed due to a subject mismatch. I'm going to extract `alpnproxy.CheckDBCerts` out of `alpnproxy.LocalProxy.CheckDBCerts` so that Connect can check the certs before actually setting them.

Does that sound good?

Reply:
Checking the certs will happen in the next PR where I actually utilize the callback, this one again is just about checking if they expired.

Reply:
This will be a bit of a pain though as I need to modify every test to create a valid cert for the gateway before we do `gateway.New`. 😩 I cannot simply do it inside `OnStart` in the middleware because from the user's perspective, it makes no sense to start a gateway with invalid certs. I'd rather fail during gateway creation and let the user reattempt the action, which will reobtain the certs.

Reply:
I'm going to do something else in the mean time in case this plan makes no sense, so I'd appreciate if you could take a look and let me know if I'm not overengineering this.

Reply:
But essentially this would entail fixing some tech debt / possible bug in Connect related to proxies that we simply weren't aware of.

Reply:
Sorry for spamming comments here. I've just realized that instead of executing the entirety of `CheckDBCerts` befor eConnect is creating a local proxy or before setting the certs, I can check for the subject mismatch alone. Essentially this is what we care about at that point (as I described in the comment with the graph above). In Connect it's not possible to start a proxy with an expired cert so the only concern would be having a cert that's for another username.

Reply:
> By always setting the cert only after we check it, we ensure that in Connect, OnNewConnection cannot encounter a cert subject mismatch.
…so Connect should treat that as unrecoverable error and not execute OnExpiredCert.

I guess my thinking is that requesting new certs from the cluster is a way to handle that error - however you're saying it's a situation that shouldn't happen, and therefore it should be "unrecoverable". I assume you mean "unrecoverable" as in, the connection just gets dropped/closed instead of trying to handle the nonsensical error. If that's what you mean, then that's fine. However you want to handle the error is ok, I was just curious why the logic was different from tsh middleware.

Reply:
> I assume you mean "unrecoverable" as in, the connection just gets dropped/closed instead of trying to handle the nonsensical error. If that's what you mean, then that's fine. 

Yep, exactly.

I'll have to reconsider this when adding per-session MFA support anyway, as I'd like to reuse a single middleware between tsh and Connect. But it doesn't seem like we'll touch per-session MFA in Connect in the next two or three months.